File: faunadb-scala/src/main/scala/faunadb/streaming/BodyValueFlowProcessor.scala
package faunadb.streaming

import java.nio.ByteBuffer
import java.util
import java.util.concurrent.{CompletableFuture, Flow, SubmissionPublisher}
import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import faunadb.QueryError
import faunadb.errors.{StreamingException, UnavailableException, UnknownException}
import faunadb.values.{StringV, VSuccess, Value}
import org.slf4j.LoggerFactory

import java.net.http.HttpResponse
import scala.collection.JavaConverters.asScalaIteratorConverter
import scala.compat.java8.FutureConverters.CompletionStageOps
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Await, Future}
import scala.util.{Failure, Success, Try}

private [faunadb] class BodyValueFlowProcessor(json: ObjectMapper, syncLastTxnTime: Long => Unit,
                                               connectionWatcher:() => CompletableFuture[HttpResponse[Flow.Publisher[java.util.List[ByteBuffer]]]] ) extends SubmissionPublisher[Value] with Flow.Processor[util.List[ByteBuffer], Value] {
  private val log = LoggerFactory.getLogger(getClass)
  private var subscription: Flow.Subscription = _
  private var subscriber: Flow.Subscriber[_ >: Value] = _

  // We do not request data from the publisher until we have one subscriber
  // to avoid discarding events before the subscriber had the chance to subscribe.
  override def subscribe(subscriber: Flow.Subscriber[_ >: Value]): Unit = {
    connectionWatcherStart(this.connectionWatcher)

    if (this.subscriber == null) {
      this.subscriber = subscriber
      super.subscribe(subscriber)
      requestOne()
    } else
      throw new IllegalStateException("BodyValueFlowProcessor can have only one subscriber")
  }

  override def onSubscribe(subscription: Flow.Subscription): Unit =
    this.subscription = subscription

  override def onNext(item: util.List[ByteBuffer]): Unit = {
    import java.nio.charset.StandardCharsets
    val text = item.iterator().asScala.map(StandardCharsets.UTF_8.decode(_).toString).mkString("")
    Try {
      val eventJsonNode = json.readTree(text)
      val value = json.treeToValue[Value](eventJsonNode, classOf[Value])

      // syncLastTxnTime if possible
      value("txn").to[Long].toOpt.foreach(syncLastTxnTime)

      // handle error in stream
      isUnrecoverableError(value, eventJsonNode) match {
        case None => submit(value)
        case Some(unrecoverableError) =>
          subscriber.onError(unrecoverableError) // notify subscriber stream
          subscription.cancel() // cancel subscription on the request body
      }
    }.recover {
      case e: Throwable =>
        log.error(s"could not handle event $text", e)
        subscriber.onError(e) // notify subscriber stream
        subscription.cancel() // cancel subscription on the request body
    }

    requestOne()
  }

  private def isUnrecoverableError(event: Value, jsonNode: JsonNode): Option[Throwable] = {
    (event("type"), Option(jsonNode.get("event"))) match {
      case (VSuccess(StringV("error"), _), Some(error)) =>
        val queryError = json.treeToValue(error, classOf[QueryError])
        val ex = new StreamingException(queryError)
        Some(ex)
      case (VSuccess(StringV("error"), _), None) =>
        Some(new UnknownException(s"unknown error received for event $event", new IllegalArgumentException()))
      case _ => None
    }
  }

  override def onError(throwable: Throwable): Unit = {
    log.error("unrecoverable error encountered by subscription", throwable)
    subscriber.onError(throwable)
  }

  override def onComplete(): Unit = {
    log.debug("subscription completed")
    subscriber.onComplete()
  }

  private def requestOne(): Unit =
    subscription.request(1)

  private def connectionWatcherStart(connectionWatcher:() => CompletableFuture[HttpResponse[Flow.Publisher[java.util.List[ByteBuffer]]]]): Unit = {
    val infiniteLoop: Future[Unit] = Future {
      while (!Thread.currentThread().isInterrupted()) {
        connectionWatcher()
          .toScala
          .flatMap {
            case successResponse if successResponse.statusCode() < 300 => Future { successResponse }
            case errorResponse => Future {onError(new Throwable(s"Status code: $errorResponse.statusCode() Body: $errorResponse.body()"))}
          }
        Thread.sleep(10000)
      }
    }

    infiniteLoop.onComplete {
      case Success(value) => println("Succeed with: " + value)
      case Failure(e) => onError(e)
    }
  }
}

File: faunadb-java/src/main/java/com/faunadb/client/streaming/BodyValueFlowProcessor.java
package com.faunadb.client.streaming;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.faunadb.client.HttpResponses;
import com.faunadb.client.errors.PermissionDeniedException;
import com.faunadb.client.types.Field;
import com.faunadb.client.types.Result;
import com.faunadb.client.types.Value;
import com.faunadb.common.Connection;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.Flow;
import java.util.concurrent.SubmissionPublisher;
import java.util.stream.Collectors;

public class BodyValueFlowProcessor extends SubmissionPublisher<Value> implements Flow.Processor<java.util.List<ByteBuffer>, Value> {

    public BodyValueFlowProcessor(ObjectMapper json, Connection connection) {
        this.json = json;
        this.connection = connection;
    }

    private static Value ErrorValue = new Value.StringV("error");
    private static Field<Long> TxnField = Field.at("txn").to(Long.class);

    private final Logger log = LoggerFactory.getLogger(getClass());
    private ObjectMapper json;
    private Connection connection;
    private Flow.Subscription subscription = null;
    private Flow.Subscriber<? super Value> subscriber = null;

    private void requestOne() {
        subscription.request(1);
    }

    @Override
    public void subscribe(Flow.Subscriber<? super Value> subscriber) {
        if (this.subscriber == null) {
            this.subscriber = subscriber;
            super.subscribe(subscriber);
            requestOne();
        } else
            throw new IllegalStateException("BodyValueFlowProcessor can have only one subscriber");
    }

    @Override
    public void onSubscribe(Flow.Subscription subscription) {
        this.subscription = subscription;
    }

    @Override
    public void onNext(List<ByteBuffer> items) {
        String text = items.stream()
            .map(b -> StandardCharsets.US_ASCII.decode(b).toString())
            .collect(Collectors.joining());

        try {
            JsonNode jsonNode = json.readTree(text);
            Value value = json.treeToValue(jsonNode, Value.class);
            // update connection last txn time
            value.getOptional(TxnField).ifPresent(ts -> connection.syncLastTxnTime(ts));

            Boolean errorEventType = value.at("type")
                .getOptional()
                .map(v -> v.equals(ErrorValue))
                .orElse(false);

            if (errorEventType) {
                HttpResponses.QueryError queryError = json.treeToValue(jsonNode.get("event"), HttpResponses.QueryError.class);
                boolean unrecoverablePermissionError = queryError.code().equals("permission denied");
                if (unrecoverablePermissionError) {
                    HttpResponses.QueryErrorResponse qer = new HttpResponses.QueryErrorResponse(401, List.of(queryError));
                    Exception ex = new PermissionDeniedException(qer);
                    subscriber.onError(ex); // notify subscriber stream
                    subscription.cancel(); // cancel subscription on the request body
                } else {
                    submit(value); // submit domain error event
                }
            } else {
                submit(value);
            }
        } catch (Exception ex){
            log.error("could not parse event " + text, ex);
            subscriber.onError(ex); // notify subscriber stream
            subscription.cancel(); // cancel subscription on the request body
        }

        requestOne();
    }

    @Override
    public void onError(Throwable throwable) {
        log.error("unrecoverable error encountered by subscription", throwable);
        subscriber.onError(throwable);
    }

    @Override
    public void onComplete() {
        log.debug("subscription completed");
        subscriber.onComplete();
    }
}
